### PR TITLE
get tests work for multiple GPUs and up to 4

### PR DESCRIPTION
**Pull request by lijinf2**
Also print out helpful message when more than 4 GPUs are detected.

### CODE REVIEW

**Comment by lijinf2** [comment]
build

File: python/run_test.sh [hunk 1]
```diff
@@ -3,6 +3,12 @@ unset SPARK_HOME
 
 python ../ci/lint_python.py --format --type-check || exit 1
 
+total_num_gpus=```nvidia-smi --query-gpu=index --format=csv,noheader | wc -l```
```
Comment (leewyang) [inline python/run_test.sh:6]: I would use single back-quote, e.g.
```
total_num_gpus=`nvidia-smi --query-gpu=index --format=csv,noheader | wc -l`
```
or even better, use the parenthesis variant:
```
total_num_gpus=$(nvidia-smi --query-gpu=index --format=csv,noheader | wc -l)
```

File: python/tests/test_linear_model.py [hunk 2]
```diff
@@ -181,7 +184,7 @@ def test_linear_regression_basic(
             spark, feature_type, data_type, X, y
         )
 
-        lr = LinearRegression()
+        lr = LinearRegression(num_workers=gpu_number)
```
Comment (leewyang) [inline python/tests/test_linear_model.py:187]: does this test need a `min(gpu_number, x)` like other tests?

File: python/tests/test_random_forest.py [hunk 3]
```diff
@@ -419,6 +422,10 @@ def test_random_forest_classifier_spark_compat(
             rf.setLeafCol("leafId")
             assert rf.getLeafCol() == "leafId"
         else:
+            assert isinstance(rf, RandomForestClassifier)
+            # reduce the number of GPUs for toy dataset to avoid empty partition
+            gpu_number = 1
+            rf.num_workers = gpu_number
```
Comment (leewyang) [inline python/tests/test_random_forest.py:428]: Should probably just use `gpu_number = min(gpu_number, 1)` to be consistent with all the other changes.

File: python/run_test.sh (hunk 1 above)
Comment (eordentlich) [reply python/run_test.sh:6]: It seems this works:  `python -c 'import cupy; print(cupy.cuda.runtime.getDeviceCount())` and it respects CUDA_VISIBLE_DEVICES.

File: python/run_test.sh [hunk 4]
```diff
@@ -3,6 +3,12 @@ unset SPARK_HOME
...
+if [ ${total_num_gpus} -gt 4 ]
+then
+        echo "Tests use at most 4 GPUs. If failed, try setting CUDA_VISIBLE_DEVICES."
+            echo "Ignore the message if CUDA_VISIBLE_DEVICES has been set properly."
```
Comment (eordentlich) [inline python/run_test.sh:10]: Looks like tabs here vs spaces.   If the above cupy python command is used, the second part of this shouldn't be needed.

File: python/tests/test_linear_model.py [hunk 5]
```diff
@@ -426,14 +429,19 @@ def test_linear_regression_spark_compat(
 
 
 @pytest.mark.parametrize("params_exception", params_exception)
-def test_fail_run_on_1_col(params_exception: Tuple[Dict[str, Any], bool]) -> None:
+def test_fail_run_on_1_col(
+    gpu_number: int, params_exception: Tuple[Dict[str, Any], bool]
+) -> None:
+    # reduce the number of GPUs for toy dataset to avoid empty partition
+    gpu_number = min(gpu_number, 1)
```
Comment (eordentlich) [inline python/tests/test_linear_model.py:436]: All tests should work for the 2 gpu case, to have some coverage for distributed case.

File: python/tests/test_nearest_neighbors.py [hunk 6]
```diff
@@ -357,6 +363,9 @@ def test_lsh_spark_compat(gpu_number: int) -> None:
     from pyspark.ml.linalg import Vectors
     from pyspark.sql.functions import col
 
+    # reduce the number of GPUs for toy dataset to avoid empty partition.
+    # cuml backend requires k <= the number of rows in the smallest index partition.
+    gpu_number = min(gpu_number, 1)
```
Comment (eordentlich) [inline python/tests/test_nearest_neighbors.py:368]: same here.

File: python/tests/test_random_forest.py (hunk 3 above)
Comment (eordentlich) [reply python/tests/test_random_forest.py:428]: Same here.

**Review by eordentlich — COMMENTED** [review]
Ideally, our tests should run on 2 gpus to have some coverage for distributed use case.

File: python/run_test.sh (hunk 1 above)
Comment (lijinf2) [reply python/run_test.sh]: Revised to "python -c ...." and used parenthesis.

File: python/tests/test_random_forest.py (hunk 3 above)
Comment (lijinf2) [reply python/tests/test_random_forest.py:432]: Revised to gpu_number = min(gpu_number, 2), and increased the size of the toy dataset.

File: python/run_test.sh (hunk 4 above)
Comment (lijinf2) [reply python/run_test.sh]: done.

File: python/tests/test_nearest_neighbors.py (hunk 6 above)
Comment (lijinf2) [reply python/tests/test_nearest_neighbors.py]: Revised to use 2 GPU and increased the size of the toy dataset.

**Comment by lijinf2** [comment]
> Ideally, our tests should run on 2 gpus to have some coverage for distributed use case.

Yeah, all toy tests should support 2 GPUs now.

**Comment by lijinf2** [comment]
build

**Review by leewyang — APPROVED** [review]

**Pull request merged**